packages/@aws-cdk/cdk/lib/cloudformation/cloudformation-token.ts

```diff
@@ -1,15 +1,59 @@
+import { Construct } from "../core/construct";
 import { resolve, Token, unresolved } from "../core/tokens";
+import { Stack } from "./stack";
 
 /**
- * Base class for CloudFormation built-ins
+ * Produce a CloudFormation expression to concat two arbitrary expressions
  */
-export class CloudFormationToken extends Token {
-  public concat(left: any | undefined, right: any | undefined): Token {
-    const parts = new Array<any>();
-    if (left !== undefined) { parts.push(left); }
-    parts.push(resolve(this));
-    if (right !== undefined) { parts.push(right); }
-    return new FnJoin('', parts);
+export function cloudFormationConcat(left: any | undefined, right: any | undefined): any {
+  if (left === undefined && right === undefined) { return ''; }
+
+  const parts = new Array<any>();
+  if (left !== undefined) { parts.push(left); }
+  if (right !== undefined) { parts.push(right); }
+
+  // Some case analysis to produce minimal expressions
+  if (parts.length === 1) { return parts[0]; }
+  if (parts.length === 2 && typeof parts[0] === 'string' && typeof parts[1] === 'string') {
+    return parts[0] + parts[1];
+  }
+
+  // Otherwise return a Join intrinsic
+  return new FnJoin('', parts);
+}
+
+export class StackAwareToken extends Token {
+  public static isInstance(x: any): x is StackAwareToken {
+    return x && x._isStackAwareCloudFormationToken;
+  }
+
+  protected readonly _isStackAwareToken: boolean;
+
+  private readonly tokenStack?: Stack;
+
+  constructor(value: any, displayName?: string, anchor?: Construct) {
+      if (typeof(value) === 'function') {
+          throw new Error('StackAwareToken can only contain eager values');
+      }
+      super(value, displayName);
+      this._isStackAwareToken = true;
+
+      if (anchor !== undefined) {
+        this.tokenStack = Stack.find(anchor);
+      }
+  }
+
+  /**
+   * In a consuming context, potentially substitute this Token with a different one
+   */
+  public substituteToken(consumingStack: Stack): Token {
```

Comment:
Would really love to explore the option to generalize this such that tokens can hook into their consumers (see #1433).